src/charm_types.py

```diff
@@ -32,18 +32,25 @@ class DatasourcePostgreSQL(BaseModel):
     uri: str = Field(min_length=1, description="Database connection URI")
 
     @classmethod
-    def from_relation(cls, relation: ops.Relation) -> "DatasourcePostgreSQL":
+    def from_relation(cls, model: ops.Model, relation: ops.Relation) -> "DatasourcePostgreSQL":
         """Create a DatasourcePostgreSQL from a relation.
 
         Args:
             relation: The relation to get the data from.
+            model: The model to get the secret from.
 
         Returns:
             A DatasourcePostgreSQL instance.
         """
         relation_data = relation.data[relation.app]
         user = relation_data.get("username", "")
         password = relation_data.get("password", "")
+        secret_user = relation_data.get("secret-user", "")
+        if user == "" and secret_user != "":  # nosec
+            secret = model.get_secret(id=secret_user)
+            secret_fields = ops.Secret.get_content(secret)
```

Comment:
I am a bit lost in here. Is the data library refreshing the secret if it has changed? Should we do it ourselves?

Reply:
We're just getting the creds from the relation data here, the library manages it's thing..

Reply:
I think the proper way of doing it is like in [here](https://github.com/canonical/paas-app-charmer/blob/3109a84fbb473396fba9e5371446965fdeb31e51/paas_app_charmer/databases.py#L61). That handles secrets and not secrets

Reply:
Oh man I actually had it done that way :face_exhaling: and with Greg we talked about moving the method in the pydantic model...but looking at the dep (the fetch method in the DatabaseRequires) it doesn't make sense to pass it into the model. So I did it like in the snippet you provided (which looks great) and I added one additional unit test. Thanks!